docs/root/intro/arch_overview/http/http_connection_management.rst

```diff
@@ -151,37 +151,56 @@ previously attempted priorities.
 Internal redirects
 --------------------------
 
-Envoy supports handling 302 redirects internally, that is capturing a 302 redirect response,
-synthesizing a new request, sending it to the upstream specified by the new route match, and
-returning the redirected response as the response to the original request.
+Envoy supports handling 3xx redirects internally, that is capturing a configurable 3xx redirect
+response, synthesizing a new request, sending it to the upstream specified by the new route match,
+and returning the redirected response as the response to the original request.
 
-Internal redirects are configured via the ref:`internal redirect action
-<envoy_api_field_route.RouteAction.internal_redirect_action>` field and
-`max internal redirects <envoy_api_field_route.RouteAction.max_internal_redirects>` field in
-route configuration. When redirect handling is on, any 302 response from upstream is
-subject to the redirect being handled by Envoy.
+Internal redirects are configured via the :ref:`internal redirect policy
+<envoy_v3_api_field_config.route.v3.RouteAction.internal_redirect_policy>` field in route configuration.
+When redirect handling is on, any 3xx response from upstream, that matches
+:ref:`redirect_response_codes
+<envoy_v3_api_field_config.route.v3.InternalRedirectPolicy.redirect_response_codes>`
+is subject to the redirect being handled by Envoy.
 
 For a redirect to be handled successfully it must pass the following checks:
 
-1. Be a 302 response.
-2. Have a *location* header with a valid, fully qualified URL matching the scheme of the original request.
+1. Have a response code matching one of :ref:`redirect_response_codes
+   <envoy_v3_api_field_config.route.v3.InternalRedirectPolicy.redirect_response_codes>`.
+2. Have a *location* header with a valid, fully qualified URL matching the scheme of the original
+   request.
 3. The request must have been fully processed by Envoy.
 4. The request must not have a body.
-5. The number of previously handled internal redirect within a given downstream request does not exceed
-   `max internal redirects <envoy_api_field_route.RouteAction.max_internal_redirects>` of the route
-   that the request or redirected request is hitting.
+5. The scheme of the downstream request and the *location* header are the same or
```

Comment:
and sorry, the API confusion above was my bad docs.
Envoy did allow serving http content over tls before.  If that's what you're looking for you can just fix the docs and remove that bit of config?  If you want to actually restrict we should find a way to call out we're changing default behavior to now match the docs (which had been incorrect, mea culpa)